frontends/concrete-python/concrete/fhe/compilation/module_compiler.py

```diff
@@ -614,6 +641,45 @@ def pretty(d, indent=0):  # pragma: no cover
                 pretty(module.statistics)
 
 
+# pylint: disable=missing-class-docstring
+class TracedOutput(NamedTuple):
+    out: Output
+    val: Any
+
+
+# pylint: disable=missing-class-docstring
+class WireTracingCm:
+    def __init__(self, module):
+        self.module = module
+
+    def __enter__(self):
+        for func in self.module.functions.values():
+
+            def _trace_wires_outputs(_self, *args):
+                return tuple(TracedOutput(Output(_self, i), arg) for (i, arg) in enumerate(args))
+
+            def _trace_wires_inputs(_self, *args):
+                for i, arg in enumerate(args):
+                    if isinstance(arg, TracedOutput):
+                        self.module.composition.wires.append(Wire(arg.out, Input(_self, i)))
+                return tuple(arg.val if isinstance(arg, TracedOutput) else arg for arg in args)
+
+            func._trace_wires_inputs = types.MethodType(_trace_wires_inputs, func)
+            func._trace_wires_outputs = types.MethodType(_trace_wires_outputs, func)
```

Comment:
I think this is a very hacky way of doing it, why not add a new optional keyword argument `wire_tracer: Optional[Callable]` to `__call__`, and use it if it's not `None`?